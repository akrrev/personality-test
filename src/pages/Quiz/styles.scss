.question-count {
  font-size: 50%;
  vertical-align: top;
}
