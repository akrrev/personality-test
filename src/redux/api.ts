import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { Question, TestResultResponse, TestSubmissionRequestArg } from 'server/types.ts';

export const api = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({ baseUrl: '/api' }),
  endpoints: (builder) => ({
    getQuestions: builder.query<Question[], void>({
      query: () => `/questions`,
    }),
    getTestResult: builder.query<TestResultResponse, TestSubmissionRequestArg>({
      query: (queryArgs) => `/test-result?score=${queryArgs.score}`,
    }),
  }),
});

export const {
  useGetQuestionsQuery,
  useGetTestResultQuery,
} = api;
