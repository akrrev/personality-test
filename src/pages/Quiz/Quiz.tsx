import { useState } from 'react';
import { Link } from 'react-router-dom';
import Question from 'components/Question/Question.tsx';
import { useGetQuestionsQuery } from 'redux/api.ts';
import './styles.scss';

export default function Quiz() {
  const { data } = useGetQuestionsQuery();
  const [activeQuestion, setActiveQuestion] = useState<number>(1);
  const [score, setScore] = useState(0);

  // Updates the total score and shows the next question.
  const submitAnswer = (id: number, questionScore: number) => {
    setScore(score + questionScore);
    setActiveQuestion(id);
  }

  const quizCompleted = data && data.length < activeQuestion;
  const title = quizCompleted
    ? `You've answered all questions`
    : (
      <>
        Answer the questions <span className="question-count">({activeQuestion}/{data?.length})</span>
      </>
    );

  return (
    <article className="page">
      <h1 className='heading-2'>
        {title}
      </h1>
        {data && data.map((item) =>
          <Question
            key={item.id}
            id={item.id}
            active={item.id === activeQuestion}
            question={item.question}
            options={item.options}
            lastQuestion={item.lastQuestion}
            onAnswer={(id, questionScore) => submitAnswer(id, questionScore)}
          />
        )}
        {quizCompleted && (
          <>
            <hr/>
            <div className="center-action">
              <Link to={`/result?score=${score}`} className="action result-link">See result</Link>
            </div>
          </>
        )}
    </article>
  )
}
