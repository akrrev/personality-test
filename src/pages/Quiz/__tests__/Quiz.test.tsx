import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { store } from '../../../redux/store';
import Quiz from '../Quiz';

jest.mock('../../../redux/store', () => ({
  __esModule: true,
  store: {
    getState: () => null,
    subscribe: () => null,
    dispatch: () => null,
  },
}));

jest.mock('../../../redux/api', () => ({
  __esModule: true,
  default: () => null,
  useGetQuestionsQuery: () => ({
    data: [
      {
        id: 1,
        question: 'First question',
        options: [
          {
            id: 1,
            title: 'First option',
            score: 1,
          },
          {
            id: 2,
            title: 'Second option',
            score: 2,
          }
        ],
      },
      {
        id: 2,
        question: 'Second question',
        options: [],
      },
    ],
  }),
}));

test('Renders Quiz component correctly', () => {
  const component = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Quiz />
      </BrowserRouter>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
