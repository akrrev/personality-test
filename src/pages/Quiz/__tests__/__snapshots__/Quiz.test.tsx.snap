// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Quiz component correctly 1`] = `
<article
  className="page"
>
  <h1
    className="heading-2"
  >
    Answer the questions 
    <span
      className="question-count"
    >
      (
      1
      /
      2
      )
    </span>
  </h1>
  <section
    className=" question"
  >
    <hr />
    <h2
      className="question-title"
    >
      First question
    </h2>
    <hr />
    <button
      className=" question-option"
      onClick={[Function]}
    >
      1
       – 
      First option
    </button>
    <button
      className=" question-option"
      onClick={[Function]}
    >
      2
       – 
      Second option
    </button>
    <div
      className="center-action"
    >
      <button
        className="question-submit"
        disabled={true}
        onClick={[Function]}
      >
        Submit answer
      </button>
    </div>
  </section>
  <section
    className="hidden question"
  >
    <hr />
    <h2
      className="question-title"
    >
      Second question
    </h2>
    <hr />
    <div
      className="center-action"
    >
      <button
        className="question-submit"
        disabled={true}
        onClick={[Function]}
      >
        Submit answer
      </button>
    </div>
  </section>
</article>
`;
